### PR TITLE
DUPLO-14801: SQS Delivery Delay

### PR DESCRIPTION
**Pull request by Duplo-nikv**
## Change description

> Updated DuploSQSQueue struct with DelaySeconds attribute.
resourceAwsSqsQueueRead function updated : setting delay_seconds key to DuploSQSQueue attribute field.
expandAwsSqsQueue function updated : Getting delay_second attribute value and type converstion.
resourceAwsSqsQueueUpdate function updated: Passing delay_seconds key to HasChanges.
Docs updated.

## Type of change
- [ ] Bug fix (fixes an issue)
- [ ✔︎] New feature (adds functionality)

## Related issues

> Fix [#1]() 

## Checklists
Creating sqs queue with delay_seconds resource attribute
Updating sqs queue by changing delay_seconds value
### Development

- [ ] Pull requests may not be submitted to the *master* branch (use *develop* instead) - or they will be closed.
- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ✔︎] Pull request has a descriptive title and context useful to a reviewer.
- [✔︎ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ✔︎] Pull request linked to task tracker where applicable

___

## **Type**
enhancement

___

## **Description**
- Added support for configuring the delivery delay of messages (`delay_seconds`) in AWS SQS queues.
- `delay_seconds` attribute allows postponing the delivery of new messages to consumers for a number of seconds within the range [0-900].
- Updated AWS SQS queue resource documentation and examples to reflect the addition of the `delay_seconds` attribute.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_aws_sqs_queue.go</strong><dd><code>Support for Delay Seconds in AWS SQS Queue Resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/resource_duplo_aws_sqs_queue.go

<li>Added <code>delay_seconds</code> attribute to the schema definition for AWS SQS <br>queue resource.<br> <li> Implemented setting and reading of <code>delay_seconds</code> in resource create, <br>read, and update functions.<br> <li> Added validation for <code>delay_seconds</code> to ensure it's within the allowed <br>range [0-900].

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/389/files#diff-fbed9fc1ac3cab814d50b1fa774ca5ec9555d581cee76c92f47bedb8088d934e">+14/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>aws_sqs.go</strong><dd><code>Added DelaySeconds Field to DuploSQSQueue Struct</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplosdk/aws_sqs.go

<li>Added <code>DelaySeconds</code> field to <code>DuploSQSQueue</code> struct to support new <br>attribute.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/389/files#diff-be61aca16136bb8368963614dc95250f6c37aa652303e44d0f0d3082fb344288">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.tf</strong><dd><code>Example Terraform Configuration for Delay Seconds in AWS SQS Queue</code></dd></summary>
<hr>
      
examples/resources/duplocloud_aws_sqs_queue/resource.tf

<li>Included <code>delay_seconds</code> in the example Terraform configuration for AWS <br>SQS queue.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/389/files#diff-7b5b89d06e914b6e9f19631cebfb00ceff8e76e8d1e9b60c5cd6422cc0ed2b0f">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Updated Changelog for Delay Seconds Attribute in AWS SQS Queue</code>&nbsp; </dd></summary>
<hr>
      
CHANGELOG.md

<li>Documented the addition of <code>delay_seconds</code> attribute to AWS SQS queue <br>resource.<br> <li> Updated documentation section for new attribute.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/389/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>aws_sqs_queue.md</strong><dd><code>Documentation Update for AWS SQS Queue Delay Seconds Attribute</code>&nbsp; </dd></summary>
<hr>
      
docs/resources/aws_sqs_queue.md

<li>Updated AWS SQS queue resource documentation to include <code>delay_seconds</code>.<br> <li> Provided example usage of the new <code>delay_seconds</code> attribute.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/389/files#diff-5b79c216053b10032817f834830d8c36a244bebb81a75f61940aac60999573b4">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+## 2024-02-08
+
+### Added
+- Added support for configuring the delay seconds (`delay_seconds`) for AWS SQS queues, allowing messages to be delayed up to 900 seconds.
```
Comment (joek-duplo) [inline CHANGELOG.md:4]: Please refer to the changed resource `duplo_aws_sqs_queue` in the changelog.  See earlier entries for some examples.

**Review by joek-duplo — APPROVED** [review]

**Review by joek-duplo — CHANGES_REQUESTED** [review]
see above comment about the changelog

**Comment by Duplo-nikv** [comment]
@CodiumAI-Agent /update_changelog --pr_update_changelog.push_changelog_changes=true'

**Comment by Duplo-nikv** [comment]
Changelog.md file updated 
Please re-review
commit 9b95a97b1d34549fcf118995e3c91911cc3cdf3e

**Review by joek-duplo — APPROVED** [review]

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-14801 TF: duplocloud_aws_sqs_queue: Add support for `Delivery Delay`](https://app.clickup.com/t/8655600/DUPLO-14801)

**Pull request merged**